### ISSUE TITLE
support for numpy arrays

### ISSUE THREAD

**Issue by KoomenErina** [duonl/synthpop-py#45]
This package assumes pandas dataframes as datatype for tables. Scikit-learn [writes ](https://scikit-learn.org/stable/faq.html#does-scikit-learn-work-natively-with-various-types-of-dataframes) that it has "limited support" for dataframes. It seems like numpy arrays are the default datatype for scikit-learn. This could mean that users that already know scikit learn might expect our package to work with numpy arrayas. Supporting numpy arrays might help with being compatible with scikit learn too.

**Comment by KoomenErina** [comment]
Differences between pandas dataframe and numpy arrays:
|pandas dataframe| numpy array|
|--------------------|--------------|
|[hetrogeneous datatype ](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html)(mixing different types of columns)| [homogeneous datatype ](https://numpy.org/doc/stable/glossary.html#term-homogeneous)(one datatype for entire array)|
|2D| any number of dimensions|
|`df['column_name']` works by default| `data['field_name']` [works only](https://numpy.org/doc/stable/user/basics.indexing.html#field-access) for structured arrays|
|[has data manipulation methods (aggregating, grouping, sql-like operations) and statistics methods](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html)| [has mostly](https://numpy.org/doc/stable/reference/routines.array-manipulation.html) primitive matrix routines|
| [not fully optimized ](https://towardsdatascience.com/speed-testing-pandas-vs-numpy-ffbf80070ee7/) for preformance| [more](https://towardsdatascience.com/speed-testing-pandas-vs-numpy-ffbf80070ee7/) optimized. |

**Comment by anneliesdej** [comment]
Below a list for pandas functions used in the PoC (potentially not complete). With models I mean methods:
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=nl style='font-family:Calibri;font-size:11.0pt'>


<div style='direction:ltr'>


Pandas function   (pandas as pd) | Used where in the   PoC
-- | --
pd.api.types.is_numeric_dtype | Plots, metric,   synth
pd.concat | Models, encoder,   synth
pd.Dataframe | Everywhere
pd.DataFrame.astype | Classifiers,   metric
pd.DataFrame.columns | Models, metric,   plots, encoder
pd.DataFrame.copy | Metric
pd.DataFrame.drop | Models
pd.DataFrame.equals | Metric
pd.DataFrame.fillna | Metric
pd.DataFrame.get | Plots
pd.DataFrame.groupby | Models
pd.DataFrame.isna | Plots, NA handling
pd.DataFrame.iterrows | Plots
pd.DataFrame.notna | Models
pd.DataFrame.nunique | Metric
pd.DataFrame.rename | Metric, models
pd.DataFrame.size | Metric
pd.DataFrame.sum | Plots
pd.DataFrame.value_counts | Plots, (commented   out in models)
pd.merge | encoder
pd.qcut | Metric



</div>


</body>

</html>

**Comment by KoomenErina** [comment]
The user should be able to use both pandas dataframes and numpy arrays. A pandas dataframe can not always be easily converted to a numpy array, since a pandas dataframe can be heterogeneous.

**Comment by KoomenErina** [comment]
In general it is a good idea to use the simplest datatype that fits your needs. In this situation the simplest datatypes are the ones build in Python. Since Pandas is build on Numpy, Numpy is "simpler" than Pandas. 

As for the requirements of the data types:

| Class/component| heterogeneous data| data manipulation/ statistics | large numeric computations| Simplest datatype that can be used
|--------------------|-----------------------|--------------------------------|--------------------------------|--------------------------------|
|`Synthesiser`         | ✅                            | ❌                                         |❌| Dataframe|
| `CartMethod`       |✅                             | ❌                                         |❌|Dataframe|
| `TreeClassifierMethod`  | ❌                   | ❌                                         |✅|Numpy array|
| `TreeRegressorMethod`  | ❌                 | ❌                                         |✅|Numpy array|
|`PCAEncoder`        | ❌                            | ❌                                         |✅|Numpy array|
|`ReplaceNoneWithValue` | ❌                  | ❌                                       |✅|Numpy array|
|`MissingValuePredictor` | ❌                  | ❌                                       |✅|Numpy array|
|`MeanEncoder` | ❌                                 | [❌ ](https://stackoverflow.com/questions/11084710/numpy-mean-with-condition)                                       |✅|Numpy array|
|`CopyMethod`|❌                                 | ❌                                 | ❌ /✅                                | no big difference, slight preference for numpy array|
|`SampleMethod`|❌                                 | ❌                                 | ❌ /✅                                | no big difference, slight preference for numpy array|
|spmse|✅                                               | ✅/❌                                    |✅|Dataframe for user convenience|
|plotting| ❌                                           | ❌                                           | ❌                     | Dataframe for user convenience|

**Comment by KoomenErina** [comment]
For the parts of the code that will be accepting numpy arrays as input,[ the type hint needs to be as wide as supported](https://typing.python.org/en/latest/guides/libraries.html#wide-vs-narrow-types). See also [duck typing](https://en.wikipedia.org/wiki/Duck_typing). In this case the widest possible type is [`numpy.typing.ArrayLike`](https://numpy.org/doc/stable/reference/typing.html#numpy.typing.ArrayLike). This is also the datatype that scikit-learn uses as input for `DecisionTreeClassifier.fit`

**Comment by anneliesdej** [comment]
**Design rationale**
In our Python package, we initially chose `pandas.DataFrame` as the primary input structure. This decision was motivated by two considerations: 1. Many of our target users transition from R to Python. For them, a `DataFrame` is intuitive and semantically aligned with R's `data.frame`; 2. A `pandas.DataFrame` natively supports heterogeneous column types within a single tabular structure. In contrast, a `numpy.array` is homogeneous by design; it enforces a single `dtype` across the entire array.

Given that our synthetic data generation methods must handle mixed-type data, the  `DataFrame` was a natural starting point. However, our estimators (`CartMethod` with the `TreeClassifierMethod` and `TreeRegressorMethod`) internally rely on decision trees from `scikit-learn`. Scikit-learn is fundamentally built around numpy arrays as its core computational substrate. Although many estimators accept `DataFrame` inputs, intrenally they convert data to `numpy.array` using `check_array`, `check_X_y`, and related validation utilities from `sklearn.utils.validation`.

Since we want our estimators to align as much as possible with the [sk-learn Developer's Guide](https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator), we tried to test our estimator using [check_estimator](https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html). This is used to validate that custom estimators comply with the scikit-learn API. These checks assume `numpy.array` as canonical input. Since our estimators were designed around `DataFrame` inputs, they failed the entire test suite. This made us re-evaluate our design.

The new proposal is to adopt the following architectural principle: **Externally we accept `pandas.DataFrame` objects for usability and internally the estimators operate on `numpy.array` objects**.

This means that within `CartMethod`, which does directly not inherit from `scikit-learn` (so can still have pandas.DataFrame as input), we create a string array for categorical features and a float array for numeric features. The categorical features will be encoded to also make them numeric. This way we create a fully homogeneous feature matrix of only float arrays, where the estimators can be fit on. The output of the synthesis (the data) will still be a `pandas.DataFrame`. For later versions, we could update to allow numpy array input as well.

This ensures:
- Compatibility with `check_estimator`
- Consistency with scikit-learn's design philosophy
- Improved [computational performance](https://towardsdatascience.com/speed-testing-pandas-vs-numpy-ffbf80070ee7/)
- Explicit and controlled `dtype` handling

In short: we keep `pandas.DataFrame` at the interface level for usability, but standardise on `numpy.array` within the estimators for correctness, performance, and sklearn compliance.

**Comment by anneliesdej** [comment]
Branch proposal for 45 is merged #50
